### ISSUE TITLE
`TransparentWrapper` derive unsoundness

### ISSUE THREAD

**Issue by steffahn** [Lokathor/bytemuck#170]
```rs
type Unit = ();

fn main() {
    let mut x = ();
    let w = Weird::wrap_mut(&mut x);
    dbg!(w);
}

use bytemuck::TransparentWrapper;

#[derive(TransparentWrapper, Debug)]
#[repr(transparent)]
#[transparent(Unit)]
struct Weird([usize; 10000], Unit);
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dfd44e5d999864f4a8af877256b71667)

**Comment by Lokathor** [comment]
Ah, it needs to assert that the size of the wrapped value is the size of the inner value. Somewhere it needs to assert that, not sure where.

**Comment by zachs18** [comment]
Hmm, I don't know if it's possible to check this (at compile time) while still allowing both:
1. Deriving `TransparentWrapper` for unsized and/or generic types (with nonwrapped fields).
2. Deriving `TransparentWrapper` for types with (nonwrapped) ZST fields which depend on a generic of the wrapper type.

(To be clear, deriving `TransparentWrapper` on a struct with only one field should never be an issue here, IIUC; this is solely about user-specified `#[transparent(WrappedType)]`.)

If 1. wasn't allowed, then you could check that `size_of::<Wrapper>() == size_of::<Wrapped>() && align_of...` in a `const _: () = assert!`, but you can't do that when `Wrapper`/`Wrapped` is unsized or generic (since they have no statically known size).

If 2. wasn't allowed, then you could check that for each non-wrapped field, `size_of::<NonWrapped>() == 0 && align_of::<NonWrapped>() == 1`, but you can't do that for a generic field, e.g. the `PhantomData<T>` in `#[transparent(usize)] struct Wrapper<T>(usize, PhantomData<T>);`.

---

I have a rough patch that gives a const-eval-time error, but removes the ability to derive TransparentWrapper for unsized and generic types, and so is probably not acceptable (it would need to be edited to give a better error message in those cases also): 
```rs
error[E0080]: evaluation of constant value failed
  --> src/main.rs:18:10
   |
18 | #[derive(TransparentWrapper, Debug)]
   |          ^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'TransparentWrapper must have the same size and alignment as the wrapped type', src/main.rs:18:10
   |
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0080`.
```

<details>

<summary>Patch</summary>

```diff
diff --git a/derive/src/traits.rs b/derive/src/traits.rs
index 1e6920b..0436c09 100644
--- a/derive/src/traits.rs
+++ b/derive/src/traits.rs
@@ -273,6 +273,7 @@ impl Derivable for TransparentWrapper {
   }
 
   fn asserts(input: &DeriveInput) -> Result<TokenStream> {
+    let wrapper_ty = &input.ident;
     let fields = get_struct_fields(input)?;
     let wrapped_type = match Self::get_wrapper_type(&input.attrs, &fields) {
       Some(wrapped_type) => wrapped_type.to_string(),
@@ -281,13 +282,25 @@ impl Derivable for TransparentWrapper {
     let mut wrapped_fields = fields
       .iter()
       .filter(|field| field.ty.to_token_stream().to_string() == wrapped_type);
-    if let None = wrapped_fields.next() {
-      bail!("TransparentWrapper must have one field of the wrapped type");
+    let wrapped_field = match wrapped_fields.next() {
+      Some(wrapped_field) => wrapped_field,
+      None => {
+        bail!("TransparentWrapper must have one field of the wrapped type")
+      }
     };
     if let Some(_) = wrapped_fields.next() {
       bail!("TransparentWrapper can only have one field of the wrapped type")
     } else {
-      Ok(quote!())
+      let wrapped_ty = &wrapped_field.ty;
+      Ok(quote!(
+        const _: () = {
+          ::core::assert!(
+            ::core::mem::size_of::<#wrapped_ty>() == ::core::mem::size_of::<#wrapper_ty>()
+            && ::core::mem::align_of::<#wrapped_ty>() == ::core::mem::align_of::<#wrapper_ty>(),
+            "TransparentWrapper must have the same size and alignment as the wrapped type"
+          )
+        };
+      ))
     }
   }
 ```
 
 </details>
 
---

Another idea would be to allow the user to put `unsafe` in the `#[transparent]` attribute to bypass the above checks. I.e. if the user puts `#[transparent(Type)]`, then the wrapper type must be nongeneric and sized, and the derive-macro will output the checks from the above patch, but if the user puts `#[transparent(unsafe(Type))]`, then the derive macro will not check anything (except what it already does), and it is the user's responsibilty to ensure that `Type` is actually the transparent wrapped type.

**Comment by steffahn** [comment]
So for comparison, [ref-cast](https://docs.rs/ref-cast/latest/ref_cast/) uses a hidden helper trait to check whether the type of all non-main fields are either `PhantomData<...>` or `PhantomPinned`. They also use a syntactic check to automatically not choose fields whose type is _called_ `PhantomData` or `PhantomPinned`, but this is merely for convenience, since for soundness one cannot rely on the apparent name of a type.

**Comment by zachs18** [comment]
Using a helper trait like @steffahn mentioned, I got it working with an error message like 
```rs
error[E0277]: the trait bound `[usize; 10000]: OneAlignmentZST` is not satisfied
  --> src/main.rs:21:14
   |
21 | struct Weird([usize; 10000], Unit);
   |              ^^^^^^^^^^^^^^ the trait `OneAlignmentZST` is not implemented for `[usize; 10000]`
   |
   = help: the following other types implement trait `OneAlignmentZST`:
             ()
             PhantomData<T>
             PhantomPinned
note: required by a bound in `assert_onezst`
  --> src/main.rs:18:10
   |
18 | #[derive(TransparentWrapper, Debug)]
   |          ^^^^^^^^^^^^^^^^^^ required by this bound in `assert_onezst`
   = note: this error originates in the derive macro `TransparentWrapper` (in Nightly builds, run with -Z macro-backtrace for more info)
```
for the code in the original issue comment, allowing only `PhantomData<T>`, `PhantomPinned`, and `()` as ZST fields, with no restrictions on the wrapped type. A "type-name-comparison convenience check"[^1] can be added.

This rough patch re-generates the `trait OneAlignmentZST` in each `const _: () = {`, once for each invocation of the `TransparentWrapper` derive macro. This could be changed to having (possibly `#[doc(hidden)]` and/or sealed) `unsafe trait OneAlignmentZST` in `bytemuck` proper, which could allow users to use custom ZSTs (if it's not sealed).

<details> <summary>Patch</summary>

```diff
diff --git a/derive/src/traits.rs b/derive/src/traits.rs
index 1e6920b..c417836 100644
--- a/derive/src/traits.rs
+++ b/derive/src/traits.rs
@@ -273,22 +273,45 @@ impl Derivable for TransparentWrapper {
   }
 
   fn asserts(input: &DeriveInput) -> Result<TokenStream> {
+    let (impl_generics, _ty_generics, where_clause) =
+      input.generics.split_for_impl();
     let fields = get_struct_fields(input)?;
     let wrapped_type = match Self::get_wrapper_type(&input.attrs, &fields) {
       Some(wrapped_type) => wrapped_type.to_string(),
       None => unreachable!(), /* other code will already reject this derive */
     };
-    let mut wrapped_fields = fields
-      .iter()
-      .filter(|field| field.ty.to_token_stream().to_string() == wrapped_type);
-    if let None = wrapped_fields.next() {
-      bail!("TransparentWrapper must have one field of the wrapped type");
-    };
-    if let Some(_) = wrapped_fields.next() {
-      bail!("TransparentWrapper can only have one field of the wrapped type")
-    } else {
-      Ok(quote!())
+    let mut found_wrapped_field = false;
+    let mut nonwrapped_field_tys = vec![];
+    for field in fields.iter() {
+      let field_ty = &field.ty;
+      if field_ty.to_token_stream().to_string() == wrapped_type {
+        if found_wrapped_field {
+          bail!(
+            "TransparentWrapper can only have one field of the wrapped type"
+          );
+        }
+        found_wrapped_field = true;
+      } else {
+        nonwrapped_field_tys.push(field_ty);
+      }
     }
+    if !found_wrapped_field {
+      bail!("TransparentWrapper must have one field of the wrapped type")
+    }
+    Ok(quote!(
+      const _: () = {
+        trait OneAlignmentZST {}
+        impl<T: ?Sized> OneAlignmentZST for ::core::marker::PhantomData<T> {}
+        impl OneAlignmentZST for ::core::marker::PhantomPinned {}
+        impl OneAlignmentZST for () {}
+        fn assert_onezst<Z: OneAlignmentZST>() {}
+        fn check #impl_generics () #where_clause {
+          #(
+            assert_onezst::<#nonwrapped_field_tys>();
+          )*
+        }
+      };
+    ))
   }
```

</details>

---

@Lokathor I'll make a PR with a refined version of this patch if you think it's acceptable, but first, which of these behaviors would be preferable?

1. Have a new `trait OneAlignmentZST` be re-generated inside a `const` at each invocation, only allow `PhantomData`/`PhantomPinned`/`()`, and have the "type-name-comparison convenience check"[^1]. (what the patch does currently (minus the name check); requires no change to `bytemuck` proper, only `bytemuck_derive`)
2. Have `pub trait OneAlignmentZST` be a sealed, possibly doc(hidden) trait in `bytemuck` proper, only allow `PhantomData`/`PhantomPinned`/`()`, and have the "type-name-comparison convenience check".
3. Have `pub unsafe trait OneAlignmentZST` be an implementable trait in `bytemuck` proper, allowing arbitrary user-defined 1-ZSTs, with no "type-name-comparison convenience check".

(1) would only require a patch version of `bytemuck_derive`; (2) and (3) would require a new patch version of `bytemuck` and (technically) a new major version of `bytemuck_derive` (see semver sidenote below).


[^1]: Like @steffahn mentioned, if only `PhantomData`/`PhantomPinned`/`()` are allowed to implement `OneAlignmentZST`, then an additional, convenience-only, non-soundness-related "type name check" could be added to the derive macro to emit better error messages. This wouldn't be compatible with having `OneAlignmentZST` be user-implementable, though.

<details> <summary> semver sidenote </summary>

*TL;DR: This is all very pedantic and probably not necessary, and can probably be entirely skipped since `bytemuck_derive` really shouldn't be used except as a dependency of `bytemuck`, and if users want to pin `bytemuck`'s version and use the derive feature, they should probably pin `bytemuck_derive` as well, so it might not be worth restricting bytemuck_derive to semver.*

(2) and (3) with only a minor (`x` in `a.x.b`) `bytemuck_derive` version bump may cause semver issues in some (dubious) cases, since newer versions of `bytemuck_derive` would emit code that doesn't work with older versions of `bytemuck` that claim to be compatible with all versions of `bytemuck_derive`  in the `1.x.y` series, but I'm not sure if that's something worth handling (I think to be affected, a dependent would have to do all three of: 1. pin `bytemuck`'s version to a version before this patch; 2. *not* pin `bytemuck_derive`'s version to a compatible version before this patch; 3. use `derive(TransparentWrapper)` on a struct with more than one field).

I haven't looked for a concrete example, but something similar may already be an issue, since every version of bytemuck since `bytemuck@1.4.0` has used a carat version requirement for `bytemuck_derive` (with major version `1`, and different minor versions), which cargo considers compatible with any later version in the `1.y.z` series, so any incompatibility between (e.g.) `bytemuck@1.4.0` (which depends on `bytemuck_derive = "^1"`) and any of `bytemuck_derive`'s `1.y.z` series and *technically* breaks cargo's semver. Even if this is true, TBH I don't think it's much of a problem (see the TL;DR above), but it can be prevented in the future by further restricting `bytemuck`'s `bytemuck_derive` dependency version requirement, or bumping `bytemuck_derive`'s major version whenever `bytemuck_derive` uses a new feature from `bytemuck`.

For future releases, it might be good to use a tilde version `bytemuck_derive = "~2.0.x"` (`>= 2.0.x, < 2.1.0`) or an exact version `bytemuck_derive = "2.0.x"` (exactly `2.0.x`) instead of carat `bytemuck_derive = "^2.0.x"` (`>= 2.0.x, < 3.0.0`), so that cargo will never select a "too new" version of `bytemuck_derive` as a dependency of `bytemuck`, and so `bytemuck_derive` can semver-compatibly emit things from newer bytemuck versions without a major version bump. (For context, `serde` and `serde_derive` are updated entirely in "lock-step"; [`serde`](https://crates.io/crates/serde/1.0.152/dependencies) pins `serde_derive` to exactly the same version as itself.  Similarly, each version of [`clap`](https://crates.io/crates/clap/4.1.4/dependencies) pins a specific version of `clap_derive`, but not in "lock-step".) Using `~x.y.z` doesn't pin an exact version, but ensures that if e.g.`bytemuck@1.15.3` declares a dependency on `bytemuck_derive = "~2.0.1"`,  then `bytemuck_derive@2.1.0` won't be selected, but `bytemuck_derive@2.0.3` could be. Pinning an exact version of `bytemuck_derive` would also work. ([cargo book link](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility))
![screenshot of "version requirement syntax" table from The Cargo Book](https://user-images.githubusercontent.com/8355914/216226086-47653958-58a1-4121-abc2-dfbd1da9042f.png)

</details>

**Comment by steffahn** [comment]
Here’s an interesting idea: leverage the compiler itself.

Generate a (hidden/local) additional copy of the `struct` in question, where the main field of type `SomeTypeExpr` is wrapped in something like `(u8, PhantomData<SomeTypeExpr>)`; the `u8` makes sure it’s not zero-sized, and the `PhantomData` makes sure no generic arguments become unused. Then there will be a compilation error if any of the _other_ fields was neither zero-sized nor 1-aligned.

**Comment by steffahn** [comment]
(Some code example for illustration: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0872f23fa5b4355d8cf2b4ad712c3498)

**Comment by a1phyr** [comment]
This is not only a size issue. A common pattern is using a ZST value as a type-level marker that some operation is safe to do. The current implementation allows creating arbitrary ZST values, which is totally unsound.

Example with `pyo3`, which use a ZST `Python` value as a marker for the GIL:
```rust
type Unit = ();

#[derive(bytemuck::TransparentWrapper)]
#[repr(transparent)]
#[transparent(Unit)]
struct GetPython<'py>(Unit, pyo3::Python<'py>);

fn unsound_get_python_gil<'py>() -> pyo3::Python<'py> {
    GetPython::wrap(()).1
}
```

To leverage the compiler, maybe we could define a `SafeZst` trait that is only defined for `PhantomData`, `PhantomPinned` and `()` and check that every other field implements it.

**Comment by steffahn** [comment]
Good catch! Seems like restricting to `PhantomData` and `PhantomPinned` might be the most straightforward approach after all. I don’t see much value in adding `()`, as it would be completely useless anyways.

**Comment by zachs18** [comment]
Thinking about this some more, we can combine the leverage-the-compiler-itself and the ZST-marker-trait ideas for a method that causes minimal breakage and doesn't require introducing a new trait:

1. Leverage the compiler, by replacing the wrapped type with a known-non-ZST to ensure that the wrapped type is the actual wrapped type.

2. Bound the non-wrapped types on `Zeroable`. Since we know that they are all 1-ZSTs by the above check leveraging the compiler, we only need to check that they are "conjurable", and `T: Zeroable + ZST` are trivially "conjurable".

This doesn't require introducing a new trait, and additionally allows custom 1-ZSTs to be used by implementing `Zeroable` for them. (`PhantomData`, `PhantomPinned`, and `()` already implement `Zeroable`)

e.g. for the struct:
```rs
#[derive(TransparentWrapper)]
#[repr(transparent)]
#[transparent(T)]
struct Test<T: ?Sized>(T, [u8; 0], PhantomPinned, PhantomData<String>);
```
we would output the checks:
```rs
const _: () = {
   #[repr(transparent)] // the compiler will emit an error here if any of the non-wrapped fields are not 1-ZSTs
   struct AssertWrappedTypeIsWrappedType<T: ?Sized>((u8, ::core::marker::PhantomData<T>), [u8; 0], PhantomPinned, PhantomData<String>);
   fn assert_zeroable<T: Zeroable>() {};
   assert_zeroable::<[u8; 0]>;
   assert_zeroable::<PhantomPinned>;
   assert_zeroable::<PhantomData<String>>;
 };
 ```

**Comment by Lokathor** [comment]
The logic at the time was "if PhantomData acts like it holds a T, and T isn't zeroable, maybe that PhantomData shouldn't also be zeroable".

But we could add the impl.
